### PR TITLE
Fix disappearing errors when re-running dmypy check

### PR DESCRIPTION
**Pull request by meshy**
This adds a commit which fixes issue https://github.com/python/mypy/issues/9655 wherein some types of error would be lost when a file was re-processed by dmypy. Regression tests are also included.

This also fixes another error where sometimes files would not be re-processed by dmypy if the only error in the file was either "unused type ignore" or "ignore without code".

### CODE REVIEW

**Comment by meshy** [comment]
Findings: the error is [removed from the fine-grained manager on this line](https://github.com/python/mypy/blob/b6cb0ed4b2e2155eddaed0bb303631ddd04ad44b/mypy/server/update.py#L429), and isn't restored thereafter. I don't yet understand what mechanism _should_ restore it, or why it is not restored.

File: test-data/unit/daemon.test [hunk 1]
```diff
@@ -522,3 +522,20 @@ class A:
...
+== Return code: 1
+$ dmypy check -- bar.py
+bar.py:2: error: Unused "type: ignore" comment
+== Return code: 1
+
+[file foo/__init__.py]
+[file foo/empty.py]
+[file bar.py]
+from foo.empty import *
+a = 1  # type: ignore
```
Comment (meshy) [inline test-data/unit/daemon.test:541]: **Question:** I have discovered that I can reproduce this with any unused import, so long as it's from a nested module.

For example, this also fails:

```
[file bar.py]
from http.client import HTTPConnection
a = 1  # type: ignore
```

This requires fewer files to be created, but presumes the existence of a module in the standard library, which may confuse the intent of the test.

At the risk of bikeshedding, is there a preferable approach in this codebase?

Comment (AlexWaygood) [reply test-data/unit/daemon.test:541]: Most tests in mypy's test suite should ideally not make any assumptions about which modules exist in the standard library (with a few exceptions such as `typing`, `typing_extensions`, etc), since typeshed's stubs for the standard library are constantly being improved and updated.

There are some exceptions to this rule — e.g. the tests in this file are only guaranteed to work if mypy uses all of typeshed's stdlib stubs: https://github.com/python/mypy/blob/master/test-data/unit/pythoneval.test

But generally speaking, we prefer to use that approach only if it's impossible to reproduce a bug properly without assuming certain stdlib features.

Comment (meshy) [reply test-data/unit/daemon.test:541]: That makes good sense to me. I'll leave the test as it is then. Thank you :)

Comment (AlexWaygood) [reply test-data/unit/daemon.test:541]: No worries!

**Comment by meshy** [comment]
Apologies if this is a little rambly. Please ask if there is anything I can clarify.

**TLDR:** There are a couple of things going on here, and I don't have a solution that solves both yet.
- **Issue 1:** We miss some important steps when modules are re-checked by `reprocess_nodes` in `update.py`.
- **Issue 2:** We do not call `reprocess_nodes` when unused type ignore errors are the only errors in a file.

**Issue 1 details:**

On the first pass (in `build.py`), unused ignore errors are added to the `manager.errors` just after calling `type_check_second_pass`. Notice that we call `generate_unused_ignore_notes`:

https://github.com/python/mypy/blob/4365dad81a248f336ee1126acbe7f8692791ca3b/mypy/build.py#L3435-L3445

On the second pass (in `update.py`), while we do a similar "second pass" check, we do not then look for unused type ignores.

https://github.com/python/mypy/blob/4365dad81a248f336ee1126acbe7f8692791ca3b/mypy/server/update.py#L1024-L1028

Notice that the above does not call `generate_unused_ignore_notes`, and is also missing:
- `generate_ignore_without_code_notes`
- `detect_possibly_undefined_vars`
- `finish_passes`

I now have an imperfect fix. Adding this line the code-block above will cause the error to be shown...

```patch
diff --git mypy/server/update.py mypy/server/update.py
index 00b823c99..7c499afb8 100644
--- mypy/server/update.py
+++ mypy/server/update.py
@@ -1027,6 +1027,9 @@ def reprocess_nodes(
         if graph[module_id].type_checker().check_second_pass():
             more = True
 
+    graph[module_id].generate_unused_ignore_notes()
+
     if manager.options.export_types:
         manager.all_types.update(graph[module_id].type_map())
```

... but _only_ if there is already another error in the file. I assume that it would make sense to call the other missing methods too, but I don't have a failing tests for them yet (and I don't know what `finish_passes` does).

**Issue 2 details:**

The `ErrorInfo` instance used to represent an unused type ignore error does not have a `target` assigned to it. This means that when a file has no other errors, it will not be reprocessed because it is not included in `FineGrainedBuildManager.previous_targets_with_errors`, which is used to [work out which files to re-process](https://github.com/python/mypy/blob/b6cb0ed4b2e2155eddaed0bb303631ddd04ad44b/mypy/server/update.py#L278-L296).

I can work around _that_ by adding this...

```patch
diff --git mypy/errors.py mypy/errors.py
index 2c2c1e5ca..552756646 100644
--- mypy/errors.py
+++ mypy/errors.py
@@ -668,6 +668,8 @@ class Errors:
                 False,
                 False,
                 False,
+                origin=(self.file, None),
+                target=self.current_target(),
             )
             self._add_error_info(file, info)
 ```
 
 (Something similar will be required in `generate_ignore_without_code_notes()`.)
 
Unfortunately, that is incompatible with the solution to **Issue 1**, because on the second pass there is some missing state, and I get this error:

```pytb
Daemon crashed!
Traceback (most recent call last):
  File ".../mypy/dmypy_server.py", line 231, in serve
    resp = self.run_command(command, data)
  File ".../mypy/mypy/dmypy_server.py", line 278, in run_command
    ret = method(self, **data)
  File ".../mypy/mypy/dmypy_server.py", line 356, in cmd_check
    return self.check(sources, export_types, is_tty, terminal_width)
  File ".../mypy/mypy/dmypy_server.py", line 415, in check
    messages = self.fine_grained_increment_follow_imports(sources)
  File ".../mypy/mypy/dmypy_server.py", line 664, in fine_grained_increment_follow_imports
    messages = fine_grained_manager.update([], to_delete)
  File ".../mypy/mypy/server/update.py", line 288, in update
    changed_modules = propagate_changes_using_dependencies(
  File ".../mypy/mypy/server/update.py", line 881, in propagate_changes_using_dependencies
    triggered |= reprocess_nodes(manager, graph, id, nodes, deps, processed_targets)
  File ".../mypy/mypy/server/update.py", line 1030, in reprocess_nodes
    graph[module_id].generate_unused_ignore_notes()
  File ".../mypy/mypy/build.py", line 2587, in generate_unused_ignore_notes
    self.manager.errors.generate_unused_ignore_errors(self.xpath)
  File ".../mypy/mypy/errors.py", line 672, in generate_unused_ignore_errors
    target=self.current_target(),
  File ".../mypy/mypy/errors.py", line 344, in current_target
    return self.scope.current_target()
  File ".../mypy/mypy/scope.py", line 33, in current_target
    assert self.module
AssertionError
```

I'm now investigating how to ensure that `manager.errors.scope.module` is appropriately set, but if anyone can suggest a course of action there, I'd appreciate it.

File: mypy/server/update.py [hunk 2]
```diff
@@ -1027,6 +1027,10 @@ def key(node: FineGrainedDeferredNode) -> int:
         if graph[module_id].type_checker().check_second_pass():
             more = True
 
+    graph[module_id].detect_possibly_undefined_vars()
```
Comment (meshy) [inline mypy/server/update.py:1030]: **Question:** I haven't added a call to `finish_passes()` here. I'm not sure what it does... should it be here?

File: mypy/errors.py [hunk 3]
```diff
@@ -720,6 +722,8 @@ def generate_ignore_without_code_errors(
                 False,
                 False,
                 False,
+                origin=(self.file, None),
+                target=self.target_module,
```
Comment (meshy) [inline mypy/errors.py:726]: **Question 1:** I'm not sure if `origin` is required here. I noticed that the other places that passed in `target=` passed this too, so I assumed it couldn't hurt. Should I remove it?

**Question 2:** I would have liked to use `target=self.current_target()` here, similar to the other uses in the file, but as I detailed in another comment that causes an error. Is this good enough? Are we losing important context by taking this approach?

**Review by meshy — COMMENTED** [review]
I'm happy that this works in my small examples now, but I have some minor questions about my approach.

**Comment by meshy** [comment]
> What is the status of this pull request?

I'm happy with the changes in this PR now, and am awaiting feedback. I don't know what the next steps are to get this merged.

**Comment by meshy** [comment]
> I keep running into this issue and it means I have to keep restarting the daemon when I'm type checking my code.

Have you tried out this branch locally? Does this solve the issue for you?

**Comment by AlexWaygood** [comment]
I don't know much about dmypy or incremental mode, so I may not be the best reviewer here (though I'm still hoping to take a look soon anyway, when I have the time!). I'm also only a triager. 

Hopefully @ilevkivskyi will be able to take a look, if he has the time :)

**Comment by CoolCat467** [comment]
> Have you tried out this branch locally? Does this solve the issue for you?

That's unfortunate, just tried installing this locally and it halfway works, subsequent runs without changing the file keep getting the same list of comments like expected, but then if the file changes still no comments even though there are errors.

If it's helpful, here is my list of flags for the daemon:
<details>
check_untyped_defs = true
disallow_any_generics = true
disallow_untyped_calls = true
disallow_untyped_defs = true
ignore_missing_imports = true
no_implicit_optional = true
no_implicit_reexport = true
show_column_numbers = true
show_error_codes = true
strict = true
strict_equality = true
warn_redundant_casts = true
warn_return_any = true
warn_unreachable = true
warn_unused_configs = true
warn_unused_ignores = true
</details>

**Comment by meshy** [comment]
> just tried installing this locally and it halfway works, subsequent runs without changing the file keep getting the same list of comments like expected, but then if the file changes still no comments even though there are errors.

Thanks for checking this out, and finding this new angle on the issue. I've confirmed that making changes to the file does cause the unused-ignore and untyped-ignore warnings to disappear. Interestingly, the possibly-undefined error appears to remain.

I haven't yet looked into how this is happening. Hopefully I will get some more time for this soon.

**Comment by ilevkivskyi** [comment]
@meshy I think to fix the issue reported by @CoolCat467 you will also need to change `update_module_isolated()`, currently it only calls `detect_possibly_undefined_vars()`, but not the other two.

Unfortunately it looks like it is not possible to avoid the code duplication without a bigger refactoring, so I think if it fixes the issue it should be fine for now.

**Comment by meshy** [comment]
Thank you for the pointer @ilevkivskyi, that really helped me to save a lot of time.

I've added in those checks, and the tests are all now passing for me locally. Please let me know if there's anything else I can do to help get this in.

**Review by ilevkivskyi — APPROVED** [review]
LGTM, thanks!

**Comment by meshy** [comment]
We're getting test failures from the `Tests / Test suite with py37-windows-64` test run, but I don't understand why.

```
================================== FAILURES ===================================
__________________ testUnusedTypeIgnorePreservedAfterChange ___________________
[gw1] win32 -- Python 3.7.9 D:\a\mypy\mypy\.tox\py37\Scripts\python.EXE
data: D:\a\mypy\mypy\test-data\unit\daemon.test:560:
Failed: Command 4 (dmypy check -- bar.py) did not give expected output
---------------------------- Captured stderr call -----------------------------
Expected:
  bar.py:2: error: Unused "type: ignore" comment (diff)
  == Return code: 1
Actual:
  bar.py:2: error: Invalid "type: ignore" comment  [syntax] (diff)
  == Return code: 1

Alignment of first line difference:
  E: bar.py:2: error: Unused "type: ignore" comment
  A: bar.py:2: error: Invalid "type: ignore" comment  [syntax]
                      ^
________________ testTypeIgnoreWithoutCodePreservedAfterChange ________________
[gw1] win32 -- Python 3.7.9 D:\a\mypy\mypy\.tox\py37\Scripts\python.EXE
data: D:\a\mypy\mypy\test-data\unit\daemon.test:578:
Failed: Command 4 (dmypy check -- bar.py) did not give expected output
---------------------------- Captured stderr call -----------------------------
Expected:
  bar.py:2: error: "type: ignore" comment without error code  [ignore-without-code] (diff)
  == Return code: 1
Actual:
  bar.py:2: error: Invalid "type: ignore" comment  [syntax] (diff)
  == Return code: 1

Alignment of first line difference:
  E: bar.py:2: error: "type: ignore" comment without error code  [ignore-with...
  A: bar.py:2: error: Invalid "type: ignore" comment  [syntax]...
                      ^
=========================== short test summary info ===========================
FAILED mypy/test/testdaemon.py::DaemonSuite::daemon.test::testUnusedTypeIgnorePreservedAfterChange
FAILED mypy/test/testdaemon.py::DaemonSuite::daemon.test::testTypeIgnoreWithoutCodePreservedAfterChange
==== 2 failed, 10703 passed, 289 skipped, 11 xfailed in 2471.58s (0:41:11) ====
```

I don't have Windows, so I can't reproduce this locally. Can someone  please help me with this?

**Comment by ilevkivskyi** [comment]
@meshy Most likely because `echo '' >> bar.py` does something wrong on Windows.

**Comment by AlexWaygood** [comment]
> I don't have Windows, so I can't reproduce this locally. Can someone please help me with this?

I can reproduce this locally on Windows, with Python 3.11, so it does seem to be something Windows-specific rather than py37-specific.

**Comment by AlexWaygood** [comment]
@meshy, applying this diff to your PR branch fixes the issue for me locally on Windows:

```diff
diff --git a/test-data/unit/daemon.test b/test-data/unit/daemon.test
index 34d1b2b92..869f60b4e 100644
--- a/test-data/unit/daemon.test
+++ b/test-data/unit/daemon.test
@@ -564,7 +564,7 @@ Daemon started
 $ dmypy check -- bar.py
 bar.py:2: error: Unused "type: ignore" comment
 == Return code: 1
-$ echo '' >> bar.py
+$ {python} -c "print('\n')" >> bar.py
 $ dmypy check -- bar.py
 bar.py:2: error: Unused "type: ignore" comment
 == Return code: 1
@@ -582,7 +582,7 @@ Daemon started
 $ dmypy check -- bar.py
 bar.py:2: error: "type: ignore" comment without error code  [ignore-without-code]
 == Return code: 1
-$ echo '' >> bar.py
+$ {python} -c "print('\n')" >> bar.py
 $ dmypy check -- bar.py
 bar.py:2: error: "type: ignore" comment without error code  [ignore-without-code]
 == Return code: 1
```

**Comment by AlexWaygood** [comment]
Closing and reopening to retrigger CI

**Comment by meshy** [comment]
Thank you both for your help, that seems right.

@AlexWaygood I have applied your patch, and listed you as a co-author on that commit.

**Comment by AlexWaygood** [comment]
@meshy congrats on your first mypy contribution! This was a tough first PR 😀

**Comment by meshy** [comment]
Thank you very much! :heart:

**Comment by JukkaL** [comment]
Unfortunately I'm planning to revert this (#15179), due to remaining issues that may be non-trivial to fix (see #15043). I'm happy to receive an updated PR that also includes fixes to the issues. By reverting the change, there's no pressure to get the fixes included in mypy 1.3, which should be out soon.

**Comment by ilevkivskyi** [comment]
TBH I don't understand the reason. Does this PR causes any troubles (new errors)? Before this PR the logic in update.py was clearly wrong, this PR makes it at least generally consistent.

**Comment by meshy** [comment]
@ilevkivskyi unfortunately yes, it did cause a regression: https://github.com/python/mypy/pull/15043

**Comment by ilevkivskyi** [comment]
OK, I see. I thought that was one of remaining unfixed errors.

File: mypy/server/update.py [hunk 4]
```diff
@@ -1027,6 +1029,10 @@ def key(node: FineGrainedDeferredNode) -> int:
...
+    graph[module_id].generate_unused_ignore_notes()
+    graph[module_id].generate_ignore_without_code_notes()
```
Comment (JukkaL) [inline mypy/server/update.py:1034]: I think that it may be more correct to only perform these after the end of a build increment, after we've propagated all changes. After a call of `reprocess_nodes` we might still be in the middle of processing some files, and some errors could still be not generated yet. So maybe we'd need to keep track of all the files that have processed nodes, and call these functions at the end?

@ilevkivskyi What do you think?

File: mypy/server/update.py (hunk 2 above)
Comment (JukkaL) [reply mypy/server/update.py:1032]: I don't think so, since `finish_passes` seems to only perform things related to non-fine-grained builds.

I think that there may be another issue here: `detect_possible_undefined_vars()` is somewhat expensive, as it runs over the entire file. In this function it would be better to only analyze the nodes that are being processed. For example, we could pass the set of targets to `detect_possibly_undefined_vars`, and it would only catch errors in these targets. I'm not sure if processing the entire file causes correctness issues, but plausibly it could.

File: mypy/server/update.py (hunk 4 above)
Comment (ilevkivskyi) [reply mypy/server/update.py:1034]: I don't know TBH, I thought we clear those "intermediate" errors. So I guess we may clear them for some reason? If yes, this should be fixed. But if we fixing that is hard, then yes, we can delay generating these errors.

**Comment by meshy** [comment]
I think I've now addressed the issues which lead to this being reverted, and [opened the new PR for review](https://github.com/python/mypy/pull/15043).

File: mypy/server/update.py (hunk 4 above)
Comment (meshy) [reply mypy/server/update.py:1034]: Hi both -- thank you for your comments on this.

Apologies for not addressing your points before.

I've re-opened this PR (with some minor changes to address a regression) in https://github.com/python/mypy/pull/15043, and I'd appreciate your feedback there. In particular, I'd really welcome your thoughts on what should be done to address the comments you left here.

Comment (meshy) [reply mypy/server/update.py:1034]: FWIW, I have removed these three lines from the new PR.
